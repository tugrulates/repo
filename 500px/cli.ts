// deno-lint-ignore-file no-console
/**
 * Command-line interface for the `500px` client.
 *
 * @module cli
 */

import { Command, EnumType, ValidationError } from "@cliffy/command";
import { Table } from "@cliffy/table";
import { displayVersion } from "@roka/package/version";
import { CATEGORIES, fiveHundredPx, type Photo } from "./500px.ts";

/**
 * Run the `500px` tool with the given command-line arguments.
 *
 * @param args Command-line arguments.
 * @returns The exit code of the command.
 */
export async function cli(args: string[]): Promise<number> {
  const cmd = new Command()
    .name("500px")
    .description("Interact with 500px.")
    .usage("<command> [options]")
    .version(await displayVersion())
    .action((): void => cmd.showHelp())
    .command("discover", discoverCommand())
    .command("follows", followsCommand())
    .command("photos", photosCommand());
  try {
    await cmd.parse(args);
  } catch (e: unknown) {
    if (e instanceof ValidationError) {
      cmd.showHelp();
      console.error(`‚ùå ${e.message}`);
      return 1;
    }
    const errors = (e instanceof AggregateError) ? e.errors : [e];
    for (const error of errors) {
      console.error(`‚ùå ${error.message}`);
      if (error["cause"] && error["cause"]["error"]) {
        console.error(error.cause.error);
      }
    }
    return 2;
  }
  return 0;
}

function discoverCommand() {
  // skips photos copied from VCG
  const skip = [/^\/vcg-/];
  return new Command()
    .description("Prints a list of active and high quality users on 500px.")
    .example(
      "500px discover",
      "Prints a list of users with high scored photos.",
    )
    .example("500px discover --filter food", "Finds food photographers.")
    .example(
      "500px discover --filter macro --filter animals",
      "Either category.",
    )
    .example("500px discover --json | jq", "Query users as JSON.")
    .type("category", new EnumType(Object.values(CATEGORIES).map((c) => c.opt)))
    .option(
      "--category <category:category>",
      "Categories to filter results on.",
      { collect: true },
    )
    .option("--json", "Output the list of users as JSON.")
    .action(async ({ category, json }) => {
      const categories = Object.values(CATEGORIES).filter((c) =>
        !category || category.includes(c.opt)
      );
      const client = fiveHundredPx();
      const photos = await client.forYouFeed({ categories, limit: 1000 });
      const users = photos.map((photo) => photo.photographer.canonicalPath)
        .filter((user) => !skip.some((re) => re.test(user)));
      const result = { discover: Array.from(new Set(users)) };
      if (json) console.log(JSON.stringify(result, undefined, 2));
      else result.discover.forEach((user) => console.log(`üë§ ${user}`));
    });
}

function followsCommand() {
  return new Command()
    .description("Prints follower information on 500px.")
    .example("500px follows", "Prints follow counts.")
    .example("500px follows --follows", "Follow users who follow.")
    .example("500px follows --unfollow", "Unfollow users who dont' follow.")
    .example("500px follows --follow --unfollow", "Matches both lists.")
    .example("500px follows --json", "Outputs JSON of follower information.")
    .example(
      "500px follows --json | jq",
      "Query JSON for follower information.",
    )
    .example(
      "500px follows --json | jq '.dontFollowBack[].username'",
      "List users who are followed but don't follow back.",
    )
    .example(
      "500px follows --json | jq '.notFollowingBack[].username'",
      "List users who follow but are not followed back.",
    )
    .arguments("<username:string>")
    .option("--json", "Output the follower information as JSON.")
    .action(async ({ json }, username) => {
      const client = fiveHundredPx();
      const [following, followers] = await Promise.all([
        client.following(username),
        client.followers(username),
      ]);
      const result = {
        following,
        followers,
        dontFollowBack: following.filter(({ id }) =>
          !followers.some((user) => user.id === id)
        ),
        notFollowingBack: followers.filter(({ id }) =>
          !following.some((user) => user.id === id)
        ),
      };

      if (json) console.log(JSON.stringify(result, undefined, 2));
      else {
        console.log(`üë§ Following ${result.following.length} people.`);
        console.log(`üë§ Followed by ${result.followers.length} people.`);
      }
    });
}

function photosCommand() {
  function output(photos: Photo[]) {
    new Table()
      .body(photos.map((photo) => [
        `üèûÔ∏è ${photo.name}`,
        `üìà${photo.pulse.highest}`,
        `‚ù§Ô∏è ${photo.likedByUsers.totalCount}`,
        `üëÅÔ∏è ${photo.timesViewed}`,
      ]))
      .render();
  }
  return new Command()
    .description("Prints the list of photos for a 500px user.")
    .example("500px photos <username>", "Prints the list of photos for a user.")
    .example("500px photos <username> --json | jq", "Query photos as JSON.")
    .arguments("<username:string>")
    .option("--json", "Output the photo information as JSON.")
    .action(async ({ json }, username) => {
      const client = fiveHundredPx();
      const photos = await client.photos(username);

      if (json) console.log(JSON.stringify(photos, undefined, 2));
      else output(photos);
    });
}

if (import.meta.main) Deno.exit(await cli(Deno.args));
