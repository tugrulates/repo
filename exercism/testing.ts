import { assertExists } from "@std/assert/exists";
import { pick } from "@std/collections";
import { toPascalCase } from "@std/text";
import type {
  Client,
  CompletionData,
  ConceptData,
  ConceptProgressionData,
  ExerciseData,
  FileContent,
  FilesData,
  IterationData,
  ReputationData,
  ReputationMeta,
  SolutionData,
  SubmissionData,
  TaskData,
  TestResultData,
  TestRunData,
  TrackData,
  UserData,
} from "./client.ts";

/** Options for the {@link fakeClient} function. */
export interface FakeClientOptions {
  /** Whether to validate or invalidate the authentication token. */
  validate?: boolean;
  /** The user to return from the API. */
  user?: UserData;
  /** The reputation to return from the API. */
  reputation?: ReputationMeta;
  /** The tracks to return from the API. */
  tracks?: TrackData[];
  /** The exercises to return from the API. */
  exercises?: ExerciseData[];
  /** The solutions to return from the API. */
  solutions?: SolutionData[];
  /** The iterations to return from the API. */
  iterations?: IterationData[];
  /** The files to return from the API. */
  files?: FilesData;
  /** The file contents to return from the API. */
  fileContents?: FileContent[];
}

/** Creates a fake Exercism client for testing purposes. */
export function fakeClient(options?: FakeClientOptions): Client {
  let {
    validate,
    user,
    reputation,
    tracks,
    exercises,
    solutions,
    iterations,
    files,
    fileContents,
  } = options ?? {};
  const client: Client = {
    token: {
      async validate() {
        assertExists(validate, "Validate not provided");
        await Promise.resolve();
        return validate;
      },
    },
    user: {
      async get() {
        assertExists(user, "User not provided");
        await Promise.resolve();
        return { user };
      },
      async reputation() {
        assertExists(reputation, "Reputation not provided");
        await Promise.resolve();
        return {
          results: [
            testReputationData({ uuid: "test-reputation-uuid-1" }),
            testReputationData({ uuid: "test-reputation-uuid-2" }),
          ],
          meta: reputation,
        };
      },
    },
    tracks: {
      async list() {
        assertExists(tracks, "Tracks not provided");
        await Promise.resolve();
        return { tracks };
      },
    },
    track: (_) => ({
      exercises: {
        async list() {
          assertExists(exercises, "Exercises not provided");
          await Promise.resolve();
          return { exercises, solutions: solutions ?? [] };
        },
      },
      exercise(slug) {
        return {
          async start() {
            await Promise.resolve();
            const solution = testSolutionData({
              exercise: { slug },
              status: "started",
              num_iterations: 0,
            });
            if (!solutions) solutions = [];
            solutions.push(solution);
            return { solution };
          },
        };
      },
    }),
    solution: (uuid) => ({
      async get() {
        const solution = solutions?.find((solution) => solution.uuid === uuid);
        assertExists(solution, "Solution not provided");
        await Promise.resolve();
        return { solution };
      },
      async complete() {
        const solution = solutions?.find((solution) => solution.uuid === uuid);
        assertExists(solution, "Solution not provided");
        if (solution.status === "iterated") solution.status = "completed";
        await Promise.resolve();
        return testCompletionData();
      },
      async publish() {
        const solution = solutions?.find((solution) => solution.uuid === uuid);
        assertExists(solution, "Solution not provided");
        solution.status = "published";
        await Promise.resolve();
        return { solution };
      },
      async sync() {
        await Promise.resolve();
        return { solution: testSolutionData({ uuid, status: "iterated" }) };
      },
      submissions: {
        async create() {
          await Promise.resolve();
          return { submission: testSubmissionData({ uuid }) };
        },
        async testRun() {
          await Promise.resolve();
          return { test_run: testTestRunData({ uuid }) };
        },
      },
      iterations: {
        async create() {
          const solution = solutions?.find((solution) =>
            solution.uuid === uuid
          );
          assertExists(solution, "Solution not provided");
          if (solution?.status === "started") solution.status = "iterated";
          await Promise.resolve();
          return { iteration: testIterationData({ uuid }) };
        },
        async latest() {
          const iteration = iterations?.find((iteration) =>
            iteration.uuid === uuid
          );
          assertExists(iteration, "Iteration not provided");
          await Promise.resolve();
          return { iteration };
        },
        async publish() {
          const iteration = iterations?.find((iteration) =>
            iteration.uuid === uuid
          );
          assertExists(iteration, "Iteration not provided");
          iteration.is_published = true;
          await Promise.resolve();
          return { iteration };
        },
      },
      files: {
        async list() {
          assertExists(files, "Files not provided");
          await Promise.resolve();
          return { files };
        },
        async submission() {
          assertExists(fileContents, "File contents not provided");
          await Promise.resolve();
          return { files: fileContents };
        },
        async iteration() {
          assertExists(fileContents, "File contents not provided");
          await Promise.resolve();
          return { files: fileContents };
        },
      },
    }),
  };
  return client;
}

/** Creates a dummy {@link UserData} object for testing. */
export function testUserData(data?: Partial<UserData>): UserData {
  return {
    handle: "test-user-handle",
    insiders_status: "ineligible",
    ...data,
  };
}

/** Creates a dummy {@link ReputationData} object for testing. */
export function testReputationData(
  data?: Partial<ReputationData>,
): ReputationData {
  return {
    uuid: "test-reputation-uuid",
    value: 1,
    text: "test-reputation-text",
    icon_url: "test/reputation.svg",
    internal_url: "test/reputation",
    external_url: "test/user/reputation",
    created_at: "2000-01-01T00:00:00Z",
    track: pick(testTrackData(), ["title", "icon_url"]),
    is_seen: false,
    links: {
      mark_as_seen: "test/reputation/mark_as_seen",
    },
    ...data,
  };
}

/** Creates a dummy {@link ReputationMeta} object for testing. */
export function testReputationMeta(
  data?: Partial<ReputationMeta>,
): ReputationMeta {
  return {
    current_page: 1,
    total_count: 123,
    total_pages: 12,
    total_reputation: 42,
    unseen_total: 1,
    links: {
      tokens: "test/reputation/tokens",
      mark_all_as_seen: "test/reputation/mark_all_as_seen",
    },
    ...data,
  };
}

/** Creates a dummy {@link TrackData} object for testing. */
export function testTrackData(data?: Partial<TrackData>): TrackData {
  const { slug = "test-track-slug" } = data ?? {};
  return {
    slug,
    title: toPascalCase(slug),
    course: true,
    num_concepts: 12,
    num_exercises: 42,
    web_url: `test/${slug}`,
    icon_url: `test/${slug}.svg`,
    tags: ["test-tag-1", "test-tag-2"],
    last_touched_at: "2000-01-01T00:00:00Z",
    is_new: false,
    links: {
      self: `test/track/${slug}`,
      exercises: `test/track/${slug}/exercises`,
      concepts: `test/track/${slug}/concepts`,
    },
    ...data,
  };
}

/** Creates a dummy {@link ExerciseData} object for testing. */
export function testExerciseData(data?: Partial<ExerciseData>): ExerciseData {
  const { slug = "test-exercise-slug" } = data ?? {};
  return {
    slug,
    type: "practice",
    title: toPascalCase(slug),
    icon_url: `test/${slug}.svg`,
    difficulty: "easy",
    blurb: "test-exercise-blurb",
    is_external: true,
    is_unlocked: false,
    is_recommended: false,
    links: {
      self: `test/exercise/${slug}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link ConceptData} object for testing. */
export function testConceptData(data?: Partial<ConceptData>): ConceptData {
  const { slug = "test-concept-slug" } = data ?? {};
  return {
    slug,
    name: toPascalCase(slug),
    links: {
      self: `test/concept/${slug}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link TaskData} object for testing. */
export function testTaskData(data?: Partial<TaskData>): TaskData {
  return {
    id: 1,
    title: "test-task-title",
    ...data,
  };
}

/** Creates a dummy {@link SolutionData} object for testing. */
export function testSolutionData(data?: Partial<SolutionData>): SolutionData {
  const { uuid = "test-solution-uuid", status = "started" } = data ?? {};
  const published = status === "published";
  const completed = status === "completed" || published;
  const iterated = status === "iterated" || completed;
  return {
    uuid,
    private_url: `test/solution/${uuid}`,
    public_url: `test/solution/${uuid}/public`,
    status,
    mentoring_status: "none",
    published_iteration_head_tests_status: published ? "passed" : "not_queued",
    has_notifications: false,
    num_views: 0,
    num_stars: 0,
    num_comments: 0,
    num_iterations: iterated ? 1 : 0,
    num_loc: 123,
    is_out_of_date: false,
    published_at: published ? "2000-01-01T00:00:00Z" : null,
    completed_at: completed ? "2000-01-01T00:00:00Z" : null,
    updated_at: "2000-01-01T00:00:00Z",
    last_iterated_at: iterated ? "2000-01-01T00:00:00Z" : null,
    track: pick(testTrackData(), ["slug", "title", "icon_url"]),
    exercise: pick(testExerciseData(), ["slug", "title", "icon_url"]),
    ...data,
  };
}

/** Creates a dummy {@link SubmissionData} object for testing. */
export function testSubmissionData(
  data?: Partial<SubmissionData>,
): SubmissionData {
  const { uuid = "test-submission-uuid" } = data ?? {};
  return {
    uuid,
    tests_status: "passed",
    links: {
      cancel: `test/submission/${uuid}/cancel`,
      submit: `test/submission/${uuid}/submit`,
      test_run: `test/submission/${uuid}/test_run`,
      ai_help: `test/submission/${uuid}/ai_help`,
      initial_files: `test/submission/${uuid}/initial_files`,
      last_iteration_files: `test/submission/${uuid}/last_iteration_files`,
    },
    ...data,
  };
}

/** Creates a dummy {@link TestRunData} object for testing. */
export function testTestRunData(data?: Partial<TestRunData>): TestRunData {
  const { uuid = "test-test-run-uuid" } = data ?? {};
  return {
    uuid,
    submission_uuid: "test-submission-uuid",
    version: 1,
    status: "pass",
    message: null,
    message_html: null,
    output: null,
    output_html: null,
    tasks: [testTaskData({ id: 1 }), testTaskData({ id: 2 })],
    highlightjs_language: "test-highlightjs-language",
    tests: [
      testTestResultData({ task_id: 1 }),
      testTestResultData({ task_id: 2 }),
    ],
    links: {
      self: `test/test_run/${uuid}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link TestResultData} object for testing. */
export function testTestResultData(
  data?: Partial<TestResultData>,
): TestResultData {
  return {
    name: "test-test-result-name",
    status: "pass",
    test_code: "test-test-result-test-code",
    message: "test-test-result-message",
    message_html: "test-test-result-message-html",
    expected: "test-test-result-expected",
    output: "test-test-result-output",
    output_html: "test-test-result-output-html",
    task_id: null,
    ...data,
  };
}

/** Creates a dummy {@link CompletionData} object for testing. */
export function testCompletionData(
  data?: Partial<CompletionData>,
): CompletionData {
  return {
    track: testTrackData(),
    exercise: testExerciseData(),
    unlocked_exercises: [testExerciseData({ slug: "unlocked" })],
    unlocked_concepts: [testConceptData({ slug: "unlocked" })],
    concept_progressions: [testConceptProgressionData({ slug: "progressed" })],
    ...data,
  };
}

/** Creates a dummy {@link ConceptProgressionData} object for testing. */
export function testConceptProgressionData(
  data?: Partial<ConceptProgressionData>,
): ConceptProgressionData {
  const { slug = "test-concept-slug" } = data ?? {};
  return {
    slug,
    name: toPascalCase(slug),
    from: 0,
    to: 1,
    total: 2,
    links: {
      self: `test/concept/${slug}`,
    },
    ...data,
  };
}

/** Creates a dummy {@link IterationData} object for testing. */
export function testIterationData(
  data?: Partial<IterationData>,
): IterationData {
  const { uuid = "test-solution-uuid" } = data ?? {};
  return {
    uuid,
    submission_uuid: "test-submission-uuid",
    idx: 1,
    status: null,
    num_essential_automated_comments: 0,
    num_actionable_automated_comments: 0,
    num_non_actionable_automated_comments: 0,
    num_celebratory_automated_comments: 0,
    submission_method: "api",
    created_at: "2000-01-01T00:00:00Z",
    tests_status: "passed",
    is_published: false,
    is_latest: true,
    links: {
      self: `test/iteration/${uuid}/iteration`,
      automated_feedback: `test/iteration/${uuid}/automated_feedback`,
      delete: `test/iteration/${uuid}/delete`,
      solution: `test/iteration/${uuid}`,
      test_run: `test/iteration/${uuid}/test_run`,
      files: `test/iteration/${uuid}/files`,
    },
    ...data,
  };
}

/** Creates a dummy {@link FilesData} object for testing. */
export function testFilesData(data?: Partial<FilesData>): FilesData {
  return {
    solution: ["test-solution-file.ts"],
    test: ["test-test-file.ts"],
    example: [".meta/test-example-file.ts"],
    ...data,
  };
}

/** Creates a dummy {@link FileContent} object for testing. */
export function testFileContent(data?: Partial<FileContent>): FileContent {
  return {
    filename: "test-filename.ts",
    content: "test-content",
    type: "solution",
    ...data,
  };
}
