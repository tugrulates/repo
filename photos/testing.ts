/**
 * This module provides test photos to test the `photos` package.
 *
 * @module testing
 */

import { pool } from "@roka/async/pool";
import { assertExists } from "@std/assert";
import { copy } from "@std/fs";
import { join } from "@std/path";
import { write, type WriteOptions } from "@tugrulates/photos/exif";
import { type Photo, photo } from "@tugrulates/photos/photo";

/**
 * Creates a temporary photo.
 *
 * @param options Update photo with given tags.
 * @returns A disposable photo with variants.
 */
export async function tempPhoto(
  options?: WriteOptions,
): Promise<Photo & AsyncDisposable> {
  const directory = await Deno.makeTempDir();
  assertExists(import.meta.dirname);
  await copy(
    join(import.meta.dirname, "__testdata__", "floating-around"),
    directory,
    { overwrite: true },
  );
  let result = await photo(directory);
  if (options) {
    await write(result.path, options);
    await pool(result.variants, (variant) => write(variant.path, options));
    result = await photo(directory);
  }
  return Object.assign(await photo(directory), {
    [Symbol.asyncDispose]: () => Deno.remove(directory, { recursive: true }),
  });
}
