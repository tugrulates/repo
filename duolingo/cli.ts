// deno-lint-ignore-file no-console
/**
 * Command-line interface for the `duolingo` client.
 *
 * @module cli
 */

import { Command, ValidationError } from "@cliffy/command";
import { Input, Secret } from "@cliffy/prompt";
import { Table } from "@cliffy/table";
import { pool } from "@roka/async/pool";
import { type Config, config } from "@roka/cli/config";
import { version } from "@roka/forge/version";
import { escape } from "@std/html";
import type { Duolingo, FeedCard } from "./duolingo.ts";
import { duolingo, TIERS } from "./duolingo.ts";
import { leagueEmoji, leagueUserEmoji, reactionEmoji } from "./emoji.ts";

type DuolingoConfig = { username: string; token: string };

/**
 * Run the `duolingo` tool with the given command-line arguments.
 *
 * @param args Command-line arguments.
 * @returns The exit code of the command.
 */
export async function cli(args: string[]): Promise<number> {
  const cfg = config<DuolingoConfig>();
  const { username, token } = await cfg.get();
  const cmd = new Command()
    .name("duolingo")
    .description("Interact with Duolingo.")
    .usage("--username <username> --token <token> <command> [options]")
    .version(await version({ target: true }))
    .example("duolingo --username <username> --token <token>", "Configure.")
    .example("duolingo --clear", "Clear the cached configuration.")
    .option("--clear", "Clear the cached configuration.", {
      standalone: true,
      action: () => cfg.clear(),
    })
    .globalEnv("DUOLINGO_USERNAME=<username:string>", "Username.", {
      prefix: "DUOLINGO_",
    })
    .globalOption(
      "--username <username:string>",
      "Username.",
      username ? { default: username } : {},
    )
    .globalEnv("DUOLINGO_TOKEN=<token:string>", "JWT token.", {
      prefix: "DUOLINGO_",
    })
    .globalOption(
      "--token <token:string>",
      "JWT token.",
      token ? { default: token } : {},
    )
    .help({ colors: Deno.stdout.isTerminal() })
    .noExit()
    .globalAction((options) => cfg.set(options))
    .command("feed", feedCommand(cfg))
    .command("follows", followsCommand(cfg))
    .command("league", leagueCommand(cfg));
  try {
    await cmd.parse(args);
  } catch (e: unknown) {
    if (e instanceof ValidationError) {
      cmd.showHelp();
      console.error(`‚ùå ${e.message}`);
      return 1;
    }
    const errors = (e instanceof AggregateError) ? e.errors : [e];
    for (const error of errors) {
      console.error(`‚ùå ${error.message}`);
      if (error["cause"] && error["cause"]["error"]) {
        console.error(error.cause.error);
      }
    }
    return 2;
  }
  return 0;
}

function feedCommand(cfg: Config<DuolingoConfig>) {
  function summary(card: FeedCard): string {
    function plain(html: string) {
      return escape(html)
        .replace(/&lt;.*?&gt;/g, "")
        .replace(/[\u200E-\u200F]/g, "");
    }
    return card.header
      ? plain(card.header)
      : `${card.displayName} ${plain(card.body).toLowerCase()}`;
  }
  return new Command()
    .description("Prints and interacts with the feed.")
    .example("duolingo feed", "Prints the feed.")
    .example("duolingo feed --engage", "Engages with the feed.")
    .example("duolingo feed --json | jq", "Query JSON over the feed.")
    .option("--engage", "Engage with the feed events.")
    .option("--json", "Output the feed as JSON.")
    .action(async ({ engage, json }) => {
      const client = await api(cfg);
      const followers = await client.follows.followers();
      const cards = await client.feed.get();
      async function react(card: FeedCard): Promise<boolean> {
        if (card.cardType === "FOLLOW") {
          const user = followers.find((user) => user.userId === card.userId);
          if (!user?.isFollowing && !user?.canFollow) {
            await client.users.follow(card.userId);
            return true;
          }
        } else if (
          card.cardType === "KUDOS_OFFER" ||
          card.cardType === "SHARE_SENTENCE_OFFER"
        ) {
          if (!card.reactionType) {
            await client.feed.react(card);
            return true;
          }
        }
        return false;
      }
      if (json) console.log(JSON.stringify(cards, undefined, 2));
      await pool(
        cards,
        async (card) => {
          if (!engage || await react(card)) {
            if (!json) console.log(`${reactionEmoji(card)} ${summary(card)}`);
          }
        },
        { concurrency: 1 },
      );
    });
}

function followsCommand(config: Config<DuolingoConfig>) {
  return new Command()
    .description("Prints and manages follower information on Duolingo.")
    .example("duolingo follows", "Prints follow counts.")
    .example("duolingo follows --follows", "Follow users who follow.")
    .example("duolingo follows --unfollow", "Unfollow users who dont' follow.")
    .example("duolingo follows --follow --unfollow", "Matches both lists.")
    .example("duolingo follows --json", "Outputs JSON of follower information.")
    .example(
      "duolingo follows --json | jq",
      "Query JSON for follower information.",
    )
    .example(
      "duolingo follows --json | jq '.dontFollowBack[].username'",
      "List users who are followed but don't follow back.",
    )
    .example(
      "duolingo follows --json | jq '.notFollowingBack[].username'",
      "List users who follow but are not followed back.",
    )
    .option("--follow", "Follow users who follow.")
    .option("--unfollow", "Unfollow users who don't follow.")
    .option("--json", "Output the follower information as JSON.")
    .action(async ({ follow, unfollow, json }) => {
      const client = await api(config);
      let result = await client.follows.get();

      if (follow || unfollow) {
        if (follow) {
          await pool(
            result.notFollowingBack,
            async (user) => {
              if (user.canFollow) {
                const ok = await client.users.follow(user);
                if (!json && ok) console.log(`‚úÖ Followed ${user.username}.`);
              }
            },
            { concurrency: 1 },
          );
        }
        if (unfollow) {
          await pool(
            result.dontFollowBack,
            async (user) => {
              const ok = await client.users.unfollow(user);
              if (!json && ok) console.log(`‚ùå Unfollowed ${user.username}.`);
            },
            { concurrency: 1 },
          );
        }
        result = await client.follows.get();
      }

      if (json) console.log(JSON.stringify(result, undefined, 2));
      else {
        console.log(`üë§ Following ${result.following.length} people.`);
        console.log(`üë§ Followed by ${result.followers.length} people.`);
      }
    });
}

function leagueCommand(config: Config<DuolingoConfig>) {
  return new Command()
    .description("Prints and interacts with the current Duolingo league.")
    .example("duolingo league", "Prints the league.")
    .example("duolingo league --follow", "Follows users in the league.")
    .example("duolingo league --json | jq", "Query JSON over the league.")
    .option("--follow", "Follow users in the league.")
    .option("--json", "Output the league as JSON.")
    .action(async ({ follow, json }) => {
      const client = await api(config);
      const league = await client.league.get();
      if (league) {
        if (follow) await client.league.follow(league);
        if (json) console.log(JSON.stringify(league, undefined, 2));
        else {
          const following = await client.follows.following();
          new Table()
            .header([leagueEmoji(league), `${TIERS[league.tier]} League`])
            .body(
              league.rankings.map((user, index) => [
                `${index + 1}.`,
                `${user.display_name} ${leagueUserEmoji(user)}`,
                following.find((f) => f.userId === user.user_id) ? "üë§" : "",
                `${user.score.toString()} XP`,
              ]),
            )
            .columns([{ align: "right" }, {}, {}, { align: "right" }])
            .render();
        }
      } else {
        if (json) console.log("{}");
        else console.log("üèÜ The league has not started yet.");
      }
    });
}

async function api(cfg: Config<DuolingoConfig>): Promise<Duolingo> {
  let { username, token } = await cfg.get();
  if (!username) username = await Input.prompt("Username");
  if (!token) token = await Secret.prompt("Token");
  if (!username || !token) throw new Error("Username and token are required.");
  return duolingo({ username, token });
}

if (import.meta.main) Deno.exit(await cli(Deno.args));
