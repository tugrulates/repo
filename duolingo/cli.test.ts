// deno-lint-ignore-file no-console
import { mockFetch } from "@roka/http/testing";
import { fakeConsole } from "@roka/testing/fake";
import { assertSnapshot } from "@std/testing/snapshot";
import { cli } from "./cli.ts";

const TESTS = [
  "feed",
  "feed --engage",
  "follows",
  "follows --follow",
  "follows --unfollow",
  "league",
  "league --follow",
];

for (const test of TESTS) {
  Deno.test(`duolingo ${test}`, {
    sanitizeOps: false,
    sanitizeResources: false,
  }, async (t) => {
    using console = fakeConsole();
    using fetch = mockFetch(t);
    const args = test.replace("<username>", "tugrulates").split(" ");
    await cli(args, {
      // Use ENV variables for recording, but fake credentials for replay.
      ...fetch.mode === "replay" ? { config: ":memory:" } : {},
    });
    await assertSnapshot(t, console.output({ wrap: "\n" }));
  });
}
