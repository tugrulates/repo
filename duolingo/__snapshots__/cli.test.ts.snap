export const snapshot = {};

snapshot[`duolingo feed 1`] = `
"
🎉 Regulita came back to learn spanish after 1 month!
🎉 Queenie finished top 3 on the leaderboard and was promoted to the diamond league!
🎉 sosoxchanel@ collected the march badge!
💖 Debbie finished top 3 in the diamond league 16 times!
💖 Atayvia Roberts finished #1 in the diamond league 4 times!
💖 Denver finished top 3 in the diamond league 45 times!
💖 Josa Fav finished top 3 in the diamond league 39 times!
💖 Lorena finished top 3 on the leaderboard and was promoted to the emerald league!
🎉 Arlette did more than 10 lessons in a day!
💖 Naysú finished #1 in the diamond league 6 times!
💖 GULCAN finished #1 in the diamond league 66 times!
💖 Victor Hugo finished #1 in the diamond league 7 times!
🎉 Bonnie Alexander did more than 10 lessons in a day!
💯 おさるのジョージ completed a 3400 day streak!
💖 Svitlana Nabivach finished top 3 in the diamond league 27 times!
💖 Paul BOHOLE was promoted to the obsidian league!
💖 AŽ finished top 3 on the leaderboard and was promoted to the pearl league!
💖 Abdullah Shosha finished top 3 on the leaderboard and was promoted to the gold league!
💖 Crodragon finished top 3 on the leaderboard and was promoted to the obsidian league!
💖 Geral finished top 3 on the leaderboard and was promoted to the diamond league!
💖 Yeilin Peña finished top 3 on the leaderboard and was promoted to the obsidian league!
💖 Karine finished top 3 in the diamond league!
💯 mir learned a total of 2000 words!
💖 Susi finished top 3 in the diamond league 52 times!
💖 Petinol finished #1 in the diamond league!
💪 Santa collected the march badge!
💪 Sacha collected the march badge!
💪 Josa Fav collected the march badge!
🎉 Beyza did more than 10 lessons in a day!
💯 Валаа Эльхадж completed a 700 day streak!
💖 Santa finished top 3 on the leaderboard and was promoted to the diamond league!
💖 Jasper finished top 3 on the leaderboard and was promoted to the diamond league!
🎉 Atiga did more than 10 lessons in a day!
💪 L a u r e collected the march badge!
💖 Юлия finished #1 in the diamond league 29 times!
💖 Mary finished #1 in the diamond league 2 times!
💪 Magno Henrique collected the march badge!
💪 Anémone collected the march badge!
💖 frankrae finished #1 in the diamond league 3 times!
💖 Анна finished #1 in the diamond league 18 times!
🎉 Sony did more than 10 lessons in a day!
🎉 prinsca mash did more than 10 lessons in a day!
💪 Hans Svendsen collected the march badge!
🎉 ELISEI_RO 7 did more than 10 lessons in a day!
🎉 Yeilin Peña did more than 10 lessons in a day!
🎉 mir completed 5 friends quests in a row!
🎉 mir did more than 10 lessons in a day!
🎉 Anna completed a 125 day streak!
"
`;

snapshot[`duolingo feed --engage 1`] = `
"
🎉 Regulita came back to learn spanish after 1 month!
🎉 Queenie finished top 3 on the leaderboard and was promoted to the diamond league!
🎉 sosoxchanel@ collected the march badge!
"
`;

snapshot[`duolingo follows 1`] = `
"
👤 Following 400 people.
👤 Followed by 395 people.
"
`;

snapshot[`duolingo follows --follow 1`] = `
"
👤 Following 400 people.
👤 Followed by 395 people.
"
`;

snapshot[`duolingo follows --unfollow 1`] = `
"
❌ Unfollowed Carmen227854.
❌ Unfollowed Maga292126..
❌ Unfollowed MamukaBeri2.
❌ Unfollowed ywxZ14.
❌ Unfollowed mokb_IT_security.
❌ Unfollowed GinaJara.
❌ Unfollowed JohnLikesKetchup.
❌ Unfollowed Computerte15.
❌ Unfollowed MustafaAli586337.
❌ Unfollowed Nurullah814.
❌ Unfollowed Lololo124.
❌ Unfollowed chiedum344211233.
❌ Unfollowed loan_gmsblay11.
❌ Unfollowed saraisabely__.
❌ Unfollowed riwo868.
❌ Unfollowed .Paluten.
👤 Following 400 people.
👤 Followed by 395 people.
"
`;

snapshot[`duolingo league 1`] = `
"
🖤  Grandmaster League        
 1. Bernadette          102 XP
 2. Karin Ytsma          81 XP
 3. ClaudiaB             70 XP
 4. PikaPika1312 😺      50 XP
 5. Vesela Ognyanova     45 XP
 6. misqsqs 🇫🇷           39 XP
 7. Diane Rawlinson      39 XP
 8. Catherine            35 XP
 9. Klaus                34 XP
10. Res                  34 XP
11. Luisa Scholz         30 XP
12. Anaël                29 XP
13. Voldie Nsona 🍿      27 XP
14. Mela Schwalbe 🇯🇵     26 XP
15. Tuğrul 👀            20 XP
16. Nadzia 🍿            13 XP
17. jo kibble            10 XP
18. Lucie                10 XP
"
`;

snapshot[`duolingo league --follow 1`] = `
"
🖤  Grandmaster League        
 1. Bernadette          102 XP
 2. Karin Ytsma          81 XP
 3. ClaudiaB             70 XP
 4. PikaPika1312 😺      50 XP
 5. Vesela Ognyanova     45 XP
 6. misqsqs 🇫🇷           39 XP
 7. Diane Rawlinson      39 XP
 8. Catherine            35 XP
 9. Klaus                34 XP
10. Res                  34 XP
11. Luisa Scholz         30 XP
12. Anaël                29 XP
13. Voldie Nsona 🍿      27 XP
14. Mela Schwalbe 🇯🇵     26 XP
15. Tuğrul 👀            20 XP
16. Nadzia 🍿            13 XP
17. jo kibble            10 XP
18. Lucie                10 XP
"
`;
